segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,247 @@
+     *
+     * @param minParallelism The minimum degree of parallelism desired.
+     * @param maxParallelism The maximum degree of parallelism desired.
+     * @param timeSupplier   A Supplier that returns the current time, in nanoseconds.
+     */
+    @VisibleForTesting
+    ConcurrencyManager(int minParallelism, int maxParallelism, Supplier<Long> timeSupplier) {
+        this.timeSupplier = Preconditions.checkNotNull(timeSupplier, "timeSupplier");
+        Preconditions.checkArgument(minParallelism <= maxParallelism, "minParallelism must be <= maxParallelism.");
+        this.minParallelism = minParallelism;
+        this.maxParallelism = maxParallelism;
+        this.recentTotalWrittenLength = 0;
+        this.recentWriteCount = 0;
+        resetSnapshot(this.timeSupplier.get());
+    }
+
+    //endregion
+
+    //region Operations
+
+    /**
+     * Records the fact that a write with the given length has just completed successfully.
+     *
+     * @param writeLength   The length of the write that completed.
+     * @param elapsedMillis The amount of time, in millis, that this write took to complete.
+     */
+    void writeCompleted(int writeLength, long elapsedMillis) {
+        synchronized (this) {
+            this.recentWriteCount++;
+            this.recentTotalWrittenLength += writeLength;
+            this.recentTotalLatencyMillis += elapsedMillis;
+        }
+    }
+
+    /**
+     * Gets a value indicating the current (most recent) degree of parallelism.
+     *
+     * @return The current degree of parallelism.
+     */
+    int getCurrentParallelism() {
+        synchronized (this.snapshotLock) {
+            return this.lastSnapshot.parallelism;
+        }
+    }
+
+    /**
+     * Updates the degree of parallelism (if needed), based on information collected since the last time this method was called.
+     * General Notes:
+     * * If Min and Max Parallelism are the same, this method doesn't change anything.
+     * * Only attempts to change something if enough time passed since the last call to this method that made a change.
+     * * Any change to Parallelism will be done in increments of 1 (either up or down).
+     *
+     * Important Observations:
+     * * BookKeeper (BK) write throughput and latency, for a constant write size, will increase with the degree of parallelism
+     * until a certain point, after which throughput plateaus and latency starts to rise at a higher rate.
+     * * The optimal degree of parallelism varies based on a number of factors, such as write size, cluster configuration
+     * and cluster load/network capacity.
+     *
+     * How it works:
+     * * Compares Throughput and Latency in the period that just ended with the period before (normalizes based on interval duration).
+     * * If Throughput increases, the Parallelism is increased.
+     * * If Throughput decreases, the Parallelism is decreased. An exception is if the last action we took prior to this
```

Comment:
*Typo*: `An exception is thrown...`. In general, it is preferable to use active voice rather than passive voice for technical writing. For example: `This method throws an exception...`.